implement sparse categorical crossentropy, enable unitests

roywei opened the pull request:
This PR completes the implementation of sparse categorical crossentropy

1. **Implement sparse categorical crossentropy**
2. **Fix #146 found when fixing sparse categorical accuracy, fixed all element wise operators**
 Now K.equal() can handle two variables with different shapes using broadcast operations

3. **Enabled unit tests:**
- [x] test_training: test_model_with_crossentropy_losses_channels_first
- [x]  losses_test.py: test_sparse_categorical_crossentropy
- [x] metrics_test.py: test_sparse_metrics: sparse_categorical_crossentropy
- [x]  losses_test.py: test_cce_one_hot
- [x]  losses_test.py: test_sparse_categorical_crossentropy_4d
- [x] metrics_test.py: test_sparse_metrics: sparse_categorical_accuracy

4. **Enabled/Tested end to end examples:**

- [x] examples: mnist_acgan.py

5. **Performance tests**

- [x] Performance comparison to categorical_crossentropy 
Sparse categorical crossentropy is faster than categorical crossentropy when dealing with large number of exclusive labels.  X1.78 faster with 1000 classes and 10000 samples. See detailed results in comment.

sandeep-krishnamurthy commented:
Super cool work @roywei

sandeep-krishnamurthy commented on keras/backend/mxnet_backend.py:
```diff
@@ -1614,11 +1614,14 @@ def equal(x, y):
         y = y.symbol
         scalar = True
-    if scalar:
+    if scalar and x.infer_shape() == y.infer_shape():
```
There is no need for this if..else. broadcast_equal will work for both same shape and shapes that require broadcasting.

sandeep-krishnamurthy replied:
Same for all other operators below.

sandeep-krishnamurthy commented on keras/backend/mxnet_backend.py:
```diff
@@ -2902,7 +2927,29 @@ def sparse_categorical_crossentropy(target, output, from_logits=False, axis=-1):
         Output tensor.
     """
-    raise NotImplementedError('MXNet Backend: Sparse operations are not supported yet.')
+    output_dimensions = list(range(len(int_shape(output))))
```
len(int_shape(output)) => ndim(output)?

sandeep-krishnamurthy commented on keras/backend/mxnet_backend.py:
```diff
@@ -1666,12 +1680,19 @@ def greater(x, y):
-    if scalar:
-        return KerasSymbol(mx.sym.Cast(x > y, dtype='uint8'))
     if isinstance(x, mx.sym.Symbol) and isinstance(y, mx.sym.Symbol):
+        # use broadcasting to do element-wise comparison if both x and y are mxnet symbol
```
Usually it is best practice not to have multiple returns in the same function. (hard to maintain and debug). Can you restructure? Same in other places.

roywei commented:
@sandeep-krishnamurthy  @kalyc 
PR is ready for review, addressed comments, did performance benchmark using the following script:
sparse categorical is X1.78 faster 
```
import time

import numpy as np
from keras import backend as K
from keras import losses

num_classes = 1000
num_samples = 10000
pred = np.random.dirichlet(np.ones(num_classes),size=num_samples)
true = np.random.randint(0, 10, (num_samples,))
y_pred = K.variable(pred)
y_true = K.variable(true)
start = time.time()
loss = K.eval(losses.sparse_categorical_crossentropy(y_true, y_pred))
end = time.time()
print("sparse_categorical_crossentropy result:", np.mean(loss))
print("sparse_categorical_crossentropy time:", end - start)


start = time.time()
y_true = K.one_hot(y_true, num_classes)
loss = K.eval(losses.categorical_crossentropy(y_true, y_pred))
end = time.time()
print("categorical_crossentropy result:", np.mean(loss))
print("categorical_crossentropy time:", end - start)
```
output
```
sparse_categorical_crossentropy result: 7.4769287
sparse_categorical_crossentropy time: 0.14634299278259277
categorical_crossentropy result: 7.4769287
categorical_crossentropy time: 0.2528080940246582
```

kalyc commented on keras/backend/mxnet_backend.py:
```diff
@@ -1741,9 +1784,20 @@ def less_equal(x, y):
+        out = KerasSymbol(mx.sym.Cast(mx.sym.broadcast_lesser_equal(lhs=x, rhs=y), dtype='uint8'))
+    elif scalar:
+        # directly use '<=' operator for element-wise comparison
+        out = KerasSymbol(mx.sym.Cast(x <= y, dtype='uint8'))
```
[minor] check style

kalyc replied:
[adjust spaces around '=']

roywei replied:
@kalyc usually there is no space for optional params values, refer to any other operators.

kalyc commented on keras/backend/mxnet_backend.py:
```diff
@@ -1718,9 +1750,20 @@ def less(x, y):
+        # directly use '<' operator for element-wise comparison
+        out = KerasSymbol(mx.sym.Cast(x < y, dtype='uint8'))
+    else:
+        # use numpy if x and x are all numbers or numpy arrays
```
if x & y*

kalyc commented on keras/backend/mxnet_backend.py:
```diff
@@ -1692,12 +1716,20 @@ def greater_equal(x, y):
+        out = KerasSymbol(mx.sym.Cast(x >= y, dtype='uint8'))
     else:
-        raise TypeError('MXNet Backend: The inputs are not valid for greater_equal operation.')
+        # use numpy if x and x are all numbers or numpy arrays
```
if x & y*

kalyc commented on keras/backend/mxnet_backend.py:
```diff
@@ -1666,12 +1682,20 @@ def greater(x, y):
+        out = KerasSymbol(mx.sym.Cast(x > y, dtype='uint8'))
     else:
-        raise TypeError('MXNet Backend: The inputs are not valid for greater operation.')
+        # use numpy if x and x are all numbers or numpy arrays
```
if x and y*

kalyc commented on keras/backend/mxnet_backend.py:
```diff
@@ -1614,12 +1614,20 @@ def equal(x, y):
+        out = KerasSymbol(mx.sym.Cast(x == y, dtype='uint8'))
     else:
-        raise TypeError('MXNet Backend: The inputs are not valid for equal operation.')
+        # use numpy if x and x are all numbers or numpy arrays
```
if x and y*

kalyc commented on keras/backend/mxnet_backend.py:
```diff
@@ -1640,12 +1648,20 @@ def not_equal(x, y):
+        out = KerasSymbol(mx.sym.Cast(x != y, dtype='uint8'))
     else:
-        raise TypeError('MXNet Backend: The inputs are not valid for not_equal operation.')
+        # use numpy if x and x are all numbers or numpy arrays
```
if x and y*

kalyc commented on keras/backend/mxnet_backend.py:
```diff
@@ -1666,12 +1682,20 @@ def greater(x, y):
-        return KerasSymbol(mx.sym.Cast(x > y, dtype='uint8'))
     if isinstance(x, mx.sym.Symbol) and isinstance(y, mx.sym.Symbol):
-        return KerasSymbol(mx.sym.Cast(mx.sym.broadcast_greater(lhs=x, rhs=y), dtype='uint8'))
+        # use broadcasting to do element-wise comparison if both x and y are mxnet symbol
```
please follow docstring conventions as mentioned here - https://www.python.org/dev/peps/pep-0257/#one-line-docstrings

Also the comments are not necessary per operator, you can consider adding this blurb about implementation of operators at the beginning of the file depending on data type

kalyc commented on keras/backend/mxnet_backend.py:
```diff
@@ -2902,7 +2957,29 @@ def sparse_categorical_crossentropy(target, output, from_logits=False, axis=-1):
+                'which has {} dimensions.'.format(len(int_shape(output)))))
+
+    mx_output = output.symbol
+    # scale predictions so that the class probas of each sample sum to 1
```
see comment above about writing the docstring - https://www.python.org/dev/peps/pep-0257/#one-line-docstrings

roywei replied:
@kalyc compare to element-wise comparison operators (e.g. K.equal), sparse_categorical_crossentropy  and categorical_crossentropy are more complicated, and it requires different backend to do different processing logic. It's important to place the inline comments.  Placing these comments in doc string will confuse users. See tensorflow_backend.py for reference.

kalyc replied:
Thanks for the explanation!
[minor] change probas to probabilities

kalyc commented on keras/backend/mxnet_backend.py:
```diff
@@ -2902,7 +2957,29 @@ def sparse_categorical_crossentropy(target, output, from_logits=False, axis=-1):
+    mx_output = mx.sym.clip(mx_output, a_min=epsilon(), a_max=1.0 - epsilon())
+    # For this operation, the probability of a given label is considered exclusive.
+    mx_output = mx.sym.pick(mx_output, target.symbol, axis=axis, keepdims=True)
+    mx_output = - mx.sym.log(mx_output, axis=axis)
```
what does this do? Why are you making this value negative? Why not use mx.sym here instead of `-`?

roywei replied:
Please refer to loss function equations: http://ml-cheatsheet.readthedocs.io/en/latest/loss_functions.html#cross-entropy

kalyc commented on keras/backend/mxnet_backend.py:
```diff
@@ -2902,7 +2956,29 @@ def sparse_categorical_crossentropy(target, output, from_logits=False, axis=-1):
+        mx_output = mx.sym.broadcast_div(mx_output, mx.sym.sum(mx_output,
+                                                               axis=axis,
+                                                               keepdims=True))
+    # clip to prevent NaN's and Inf's
```
'infinities'

roywei commented:
@kalyc  addressed comments.
The performance scripts is just a simple one to test the operator, it's a sanity check.
Only complete and useful models should be added to benchmark scripts.
